go/service/simplefs.go

```diff
@@ -12,6 +14,67 @@ import (
 	"golang.org/x/net/context"
 )
 
+// transporterWaiter exposes a get() method to get a transporter, and waits for
+// it if it's not available yet. Additionally, it also tries to have get()
+// calls from different goroutines return in the same order they come in. This
+// is done by assigning a number to each call, and process them one-by-one.
+// Only the current processing goroutine actively checks and waits for the
+// transporter. Other calls won't start until their previous ones have
+// finished.
+type transporterWaiter struct {
+	getter func() rpc.Transporter
+
+	// Just use atomic instead of having cond protect this, to minimize the
+	// chance that multiple calls wait on the same Mutex (which is not
+	// necessarily FIFO).
+	lastWaitingNumber uint64
+
+	cond                *sync.Cond
+	lastProcessedNumber uint64
+}
+
+func newTransporterWaiter(
+	tranporterGetter func() (transporter rpc.Transporter),
+) *transporterWaiter {
+	return &transporterWaiter{
+		getter:              tranporterGetter,
+		lastWaitingNumber:   0,
+		cond:                sync.NewCond(&sync.Mutex{}),
+		lastProcessedNumber: 0,
+	}
+}
+
+const waitForTransporterInterval = time.Second / 4
+const waitForTransporterTimeout = time.Minute
+
+func (w *transporterWaiter) get(ctx context.Context) (xp rpc.Transporter) {
+	myNumber := atomic.AddUint64(&w.lastWaitingNumber, 1)
+
+	ctx, cancel := context.WithTimeout(ctx, waitForTransporterTimeout)
+	defer cancel()
+
+	w.cond.L.Lock()
+	defer w.cond.L.Unlock()
+	for w.lastProcessedNumber != myNumber-1 {
+		w.cond.Wait()
```

Comment:
Or actually, just have each goroutine wait independently for the xp.  Is there really an advantage to making them fetch it one at a time?

Reply:
If the context for a previous call is canceled, they do a `.Broadcast()` call as well so this just causes a 0.25s delay call, which can be eliminated by checking `ctx.Done()` before calling `getter()`, so that `.Sleep()` won't happen. Then all goroutines that are waiting for their number to be called will just finish quickly one by one. Does this sound like an acceptable solution?

I think using a queue or heap works, but they seem to cause more overhead than this and I was hoping to have a lightweight system here.

The FIFO is something nice to have but I don't think anywhere in GUI depends on that. I just feel like the RPC mechanism should deliver those calls in the same order they were received.

Reply:
That doesn't sound like a good solution because the first waiter could be waiting for minutes without canceling, and the second one could have canceled right away and will never see a broadcast.  Or am I missing something?

If you're looking for simple, I think you should abandon FIFO and have every do the .25s loop independently.  A real RPC system would have no guarantees on the timing of individual, independent RPCs from different goroutines.

Reply:
> the first waiter could be waiting for minutes without canceling

That doesn't sound right; the first waiter (or any waiter at the front of the line, i.e. with a waitingNumber being last processed plus one) will never wait.

> the second one could have canceled right away and will never see a broadcast

The second one doesn't return if ctx is done, but instead it just skips the sleep, increament the processed number, and call broadcast to wake up the next one in line.

That said, it sounds like this is gonna be tough to understand and debug in the future, so perhaps I should abandon it and do the independent wait like you suggested. (Still, please let me know if above makes sense.)

Reply:
> That doesn't sound right; the first waiter (or any waiter at the front of the line, i.e. with a waitingNumber being last processed plus one) will never wait.

Sorry, by "first waiter" I actually meant the one that's looping on `getter`, waiting for `xp`.

> The second one doesn't return if ctx is done, but instead it just skips the sleep, increament the processed number, and call broadcast to wake up the next one in line.

Sorry, by "sleep" I actually meant the `cond.Wait()`.  If one goroutine is looping forever on `xp` without canceling their context, and the goroutine waiting on cond will never return even if they have canceled their context.

Reply:
Ah, I see what you are saying now. The second call could have a shorter timeout than the previous ones, and this would end up "extending" their timeout to be the same as the longest one before it. Makes sense. Thanks